Update login screen for existing logged in user

razvanpapadopol opened the pull request:
#### Changes proposed in this Pull Request

* increase the gravatar image size and remove login form and other links from initial view
* display `Log in with another account` text below with a link to show login form and everything else

<img width="1099" alt="image" src="https://user-images.githubusercontent.com/87168/76763481-dfc45800-679b-11ea-9f91-dca08bca44f3.png">


#### Testing instructions
- Stay logged-in 
- Open `/log-in`

Fixes (together with https://github.com/Automattic/wp-calypso/pull/40314): #39988 
Related: #35309

Commit:
Update login screen for existing logged in user.
- add footer prop to Login block
- use showContinueAsUser to render ContinueAsUser component
- add continueAsAnotherUser to Login block state

Commit:
Update ContinueAsUser style and link href

simison commented on client/blocks/login/continue-as-user.jsx:
```diff
@@ -50,23 +51,37 @@ function ContinueAsUser( { currentUser, redirectUrlFromQuery } ) {
-				components: { userName: redirectLink },
-				comment: 'Alternative link under login header, skips login to continue as current user.',
-			} ) }
+			<Gravatar user={ currentUser } imgSize={ 200 } size={ 200 } />
```
Can we use imgSize `400` with `size` 200 to ensure this renders nicely on retina?

simison commented on client/blocks/login/continue-as-user.jsx:
```diff
@@ -50,23 +51,37 @@ function ContinueAsUser( { currentUser, redirectUrlFromQuery } ) {
+					components: {
+						link: (
+							<a
+								href="#/"
```
Does [`Button`](https://wpcalypso.wordpress.com/devdocs/design/button) component have any state that would look like a link but not need `href`?

tyxla replied:
I guess this should just be a `<button />` if there's no need for a `href`.

razvanpapadopol replied:
Yes, that's true. Using `<a/>` here is already not good for a11y because clicking it doesn't change the URL. There is some `is-link` class for `@automattic/components` `<Button/>` but it seems to be [deprecated](https://github.com/Automattic/wp-calypso/blob/master/packages/components/src/button/style.scss#L298) long time ago and it's  also causing alignment issues if used.

razvanpapadopol replied:
Fixed in 61539507267ce33c43d23e525a06dd3626a23985

simison reviewed (approved):
Tests well. Tested:

- connecting Jetpack, no signup form presented when already logged in
- open `/log-in` and `/log-in/en` as authenticated, non-authenticated, and 2fa users
- open `/log-in?site=SITE_SLUG` same as above
- Tested that actually logging in with another user in this scenario worked

With `?site=SITE_SLUG` still works (my name is the site name in this case):

<img width="1056" alt="image" src="https://user-images.githubusercontent.com/87168/76763364-a68be800-679b-11ea-84c2-cbe41f7807e8.png">

Non-2FA user, no first+lastname set, english locale:

<img width="1051" alt="image" src="https://user-images.githubusercontent.com/87168/76763328-93791800-679b-11ea-9203-f8d0b09c8564.png">

With 2FA user in different locale:

<img width="1099" alt="image" src="https://user-images.githubusercontent.com/87168/76763481-dfc45800-679b-11ea-9f91-dca08bca44f3.png">

(These texts just need translating)

simison commented:
Added ` [Status] String Freeze` label to ensure these are sent to translation before we merge this.

simison commented:
There's a short gap when the login form is rendered while data loads and this avatar step isn't loaded:

https://cloudup.com/cL5OLeTemOW

Assuming that would be the case even if we would constantly show the form, this isn't a big change in behavior.

To prevent this, we should hide the login form until API response tells us that the user in-fact is logged in. That would delay the login form visibility for many, while this visual jump is visible to few, so I think it's acceptable.

@Automattic/team-calypso @sgomes  would you like to review/test this before merging as it's touching the login landingpage?

sgomes commented:
> @Automattic/team-calypso @sgomes would you like to review/test this before merging as it's touching the login landingpage?

Thanks, @simison! I think this PR mostly deals with UI questions, not performance considerations, so I'm going to leave any considerations to more knowledgeable folks in that area. There seems to be nothing concerning from a brief glance at the code and the ICFY results.

simison commented:
@sgomes sure, no worries 👍 Just pinged you because you introduced the initial feature in https://github.com/Automattic/wp-calypso/pull/35309 and wanted to make sure I'm not missing anything. :-)

sgomes commented:
> @sgomes sure, no worries 👍 Just pinged you because you introduced the initial feature in #35309 and wanted to make sure I'm not missing anything. :-)

Thank you for keeping me in the loop, @simision! 🙂 It's nice to see folks taking the idea and improving it further 🎉

simison commented:
Some of the test failures seem related (test runner cannot see an element with `#usernameOrEmail`) and might need adjusting before the merge.

amamujee commented:
Not sure if it's something weird with my browser but some of the spacing looks off and the old login screen appeared for a second before the new shiny one. 
<img width="392" alt="Log_In_—_WordPress_com" src="https://user-images.githubusercontent.com/1103209/76782360-3a23df80-67c1-11ea-9d43-9ca1c48191e4.png">

lancewillett commented:
Quick test in Firefox latest stable on macOS, using a test user (non-Automattician).

Before | After
|------------ | -------------|
| <img src="https://user-images.githubusercontent.com/66797/76786794-0fac3500-6775-11ea-85ba-d37c4b002b35.png" width="480" /> | <img src="https://user-images.githubusercontent.com/66797/76786806-15097f80-6775-11ea-9a82-8f7f76b3d8ec.png" width="480" />

razvanpapadopol commented:
Thanks for checking this out @amamujee !

> some of the spacing looks off 

Looks like the new CSS definitions included in this PR are ignored by your browser. I'm not sure if there can be a caching issue given the hash in the live URL. Anyway if this doesn't go away with a hard refresh (CMD + shift + R in Chrome), please share the browser version. I've tested in latest Chrome/Firefox/Safari including mobile and it looks fine.

> old login screen appeared for a second before the new shiny one.

This is the UI issue @simison [mentioned earlier](https://github.com/Automattic/wp-calypso/pull/40125#issuecomment-599570623) and it's illustrated in this demo: https://cloudup.com/cL5OLeTemOW
We need to decide if the jump is acceptable from UX perspective because hiding the login form until user data is available would cause a delay (1 second or more depending on the bandwidth) in showing the whole login block. cc @Automattic/dotcom-create-design

simison commented:
>We need to decide if the jump is acceptable from UX perspective because hiding the login form until user data is available would cause a delay (1 second or more depending on the bandwidth) in showing the whole login block.

We could ease it a bit by fading the avatar into the view, so it'll look more intentional instead of the sudden jump.

lancewillett commented:
Comment on the idea to improve the jump. Let's ship without that for now. Because this change was meant to be a quick adjustment to a much-used screen, speed is more important than perfect.

The change is already a major improvement for users, and those additional improvements fall into the nice-to-have category, "maybe later."

amamujee commented:
How hard is the fade to implement? 

When I tried it, I personally felt it was pretty jarring and also like there was a bug or screen that I missed that made me feel anxious. If the fade is not easy, I think the delay is a better UX than the disappearing screen for 1 second.

I think @razvanpapadopol and @simison  should make the call here though.

Commit:
Double gravatar image size to support retina.

tyxla commented on client/blocks/login/continue-as-user.jsx:
```diff
@@ -50,23 +51,37 @@ function ContinueAsUser( { currentUser, redirectUrlFromQuery } ) {
+						),
+					},
+					args: { userName },
+					comment: 'Link to continue login as different user ',
```
```suggestion
					comment: 'Link to continue login as different user',
```

tyxla commented on client/blocks/login/index.jsx:
```diff
@@ -392,6 +415,17 @@ class Login extends Component {
+			// someone is already logged in, offer to proceed to the app without a new login
+			return (
+				<ContinueAsUser
+					onChangeAccount={ () => {
```
Should this be declared as a method instead of being redefined on each render?

razvanpapadopol replied:
Yes. Thanks for pointing this out! Fixed in https://github.com/Automattic/wp-calypso/pull/40125/commits/3882399e392ca7daec22ab9dac4b07b82bdf3ec3

Commit:
Update client/blocks/login/continue-as-user.jsx

remove empty space

Co-Authored-By: Marin Atanasov <8436925+tyxla@users.noreply.github.com>

Commit:
Change link without href to button.

Commit:
- Add handleContinueAsAnotherUser method to Login
- Remove handleChangeAccount const from ContinueAsUser

Commit:
Update login-page e2e lib to fix failing tests.
- add loginAsAnotherUser id to ContinueAsUser component
- add try/catch block in login method of LoginPage

razvanpapadopol commented:
> Some of the test failures seem related (test runner cannot see an element with `#usernameOrEmail`) and might need adjusting before the merge.

Yes, some of the tests were expecting the login form to always be there. Fixed in https://github.com/Automattic/wp-calypso/pull/40125/commits/898153be7c3534fedae868aa819c262690816a80

cc @Automattic/e2e-test-reviewers

lancewillett commented:
@razvanpapadopol What do you think about @simison's suggestion to hide the login form until the API request is successful? 

Now that I understand the issue, I can see how the login form hide / show delay even on a fast connection is not ideal.

Another suggestion is to use a loading icon and then once ready, show the correct version of the page.

Commit:
Add fade-in effect to ContinueAsUser.

razvanpapadopol commented:
@lancewillett Just added a fade-in effect to smooth a bit the transition to the "Continue as user" view. It looks like this: https://cloudup.com/cHw4ezRWvYO

The current implementation is leveraging server-side rendering to display the login form in a disabled state straight away. It's more meaningful than a loading spinner and it's not interactive so we can consider it a placeholder that is already giving some context to most of the users that are not already logged in when coming to this page. So my personal opinion is to use the fade (which we can further fine tune if needed) in order to keep the current perceived loading time for our users.

Commit:
Update LoginPage in e2e tests lib.
- don't wait for #loginAsAnotherUser

bsessions85 commented on test/e2e/lib/pages/login-page.js:
```diff
@@ -24,10 +24,19 @@ export default class LoginPage extends AsyncBaseContainer {
+		try {
+			// If there is already a logged in user, we display the Gravatar and 'Continue as [user]' button and hide login form
+			await driverHelper.setWhenSettable( driver, userNameSelector, username );
+		} catch {
```
Lets move this wait in to the try below. If the userNameSelector isn't there, the test will fail.

bsessions85 reviewed (dismissed):
I added on change request. I don't know why the test failed in CI, as it passes locally. Try this change and if it doesn't work I can get on the CI server and troubleshoot it there.

Commit:
Move check for userNameSelector to be present inside try/catch block

Commit:
cleanup unused CSS selector

Commit:
Update ContinueAsUser to use existing translated strings.

simison reviewed (approved):
Looks good and works well. 👍 

Once translations are done, we can switch to final texts in a separate PR.

Thanks for tackling this one!

Commit:
Update change user link text

razvanpapadopol commented:
> Once translations are done, we can switch to final texts in a separate PR.

Merging this one now and once translations are ready we can do the switch on https://github.com/Automattic/wp-calypso/pull/40314 🚀

lancewillett commented:
🎉 Testing it live on iPhone XS with mobile Safari — smooth transition and works as expected.

![IMG_23F23985D6DC-1](https://user-images.githubusercontent.com/66797/77183707-0203f180-6a8c-11ea-8792-46b47fd23a14.jpeg)

olaolusoga commented:
Late design feedback, it'd be great to get these changes in before pushing:

After looking at `Non-2FA user, no first+lastname set, english locale:` screenshot, made small adjustments to address translation, and longer usernames.

Design feedback:

- Bump down the gravatar width/height to 110px—it's currently at 200px.
- Add 1px border to gravatar. Border color: #A7AAAD
- Gravatar should be clickable—it isn't at the moment.
- Can we make sure everything is center.
- Remove username from button, and add below gravatar image (see mock)
- Update copy from "Not [username]? Login in with a different account" to "Not you?
Log in with another account" (see screenshot)

Updates 
Only first name
![short username](https://user-images.githubusercontent.com/1382943/77357426-84a3df80-6d1e-11ea-859c-e5a17aa2d735.png)


First and last name
![longer username](https://user-images.githubusercontent.com/1382943/77357422-82418580-6d1e-11ea-98a5-aeb8eabe71c7.png)


no first and last name, no set gravatar image
![long username, no image](https://user-images.githubusercontent.com/1382943/77357417-7eadfe80-6d1e-11ea-9551-ae7fd722cf1e.png)

Updates map closely to mobile view. **Here's mobile view mock** (for reference on things to adjust):
![X - 1](https://user-images.githubusercontent.com/1382943/77357761-10b60700-6d1f-11ea-835f-5ec9204771e5.png)

lancewillett commented:
Work still needed to wrap up the feedback: https://github.com/Automattic/wp-calypso/pull/40314

roo2 commented:
created #40556 to capture the design feedback
